### PR TITLE
Fix `hold1`

### PR DESCRIPTION
**Pull request by armanbilge**
Fixes https://github.com/typelevel/fs2/issues/3081.

### CODE REVIEW

**Review by Jasper-M — APPROVED** [review]
Still a shame that `uncons` and `concurrently` don't seem to be composable.

**Comment by mpilquist** [comment]
> Still a shame that uncons and concurrently don't seem to be composable.

I wouldn't necessarily conclude that. There's some bug hiding here that broke `hold1` but I'd be surprised if that bug was something fundamental about `uncons` and `concurrently` interactions.

**Comment by armanbilge** [comment]
> I'd be surprised if that bug was something fundamental about `uncons` and `concurrently` interactions.

Yes, it needed the `zip` as well.

I don't _really_ understand how all this scoping works, but it seemed like what was happening was some scope was being opened in the `uncons`, and because the tail was being run elsewhere `concurrently`, it was unable to locate this already open scope. If my understanding is right, then we are lacking some mechanism to thread that opened scope through.

**Comment by armanbilge** [comment]
Wait, maybe there is a way :) let me try something

https://github.com/typelevel/fs2/blob/597102bcad333c43c7fd9a413d1e16fdbe6208eb/core/shared/src/main/scala/fs2/Pull.scala#L460-L467

**Comment by armanbilge** [comment]
No, that didn't work either. be8d3e741e06b9f156e0bc2850262cf342caa7ce

**Comment by armanbilge** [comment]
Yeah, so `extendScopeTo` extended the _lifetime_ of the scope, but it still didn't use that scope to compile the `concurrent` stream.

**Comment by armanbilge** [comment]
I experimented with the idea that the scope of the background stream in `concurrently` should be part of the scope tree of the foreground stream, instead of creating its own root scope. I believe that would solve this issue, without requiring to change the implementation of `hold1`.

I tried two different approaches, either interfacing directly with `Pull.compile` or adding a new API `stream.compileScope(scope)`.

- https://github.com/typelevel/fs2/commit/779a1709fd74db8856cf5c296f81e9eceb0a5367 fixes `hold1`, but breaks `concurrently`
- https://github.com/typelevel/fs2/commit/576acc46495e8d9f6a8954da21465fff68ec1785 didn't fix `hold1`, but didn't break `concurrently`

Once I've mustered up new courage I'll probably take another shot at it.

File: core/shared/src/main/scala/fs2/Stream.scala [hunk 1]
```diff
@@ -1541,12 +1541,21 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
...
+      case Some((o, tail)) =>
+        Pull.eval(SignallingRef.of[F2, O2](o).flatTap(signal.complete(_))).flatMap { ref =>
+          tail.foreach(ref.set(_)).underlying
+        }
+
+      case None => Pull.done
     }.streamNoScope
 
+    Stream.eval(Deferred[F2, Signal[F2, O2]]).flatMap { signal =>
+      Stream.eval(signal.get).concurrently(go(signal))
```
Comment (Jasper-M) [inline core/shared/src/main/scala/fs2/Stream.scala:1555]: This will result in `Stream.never` if the signal never gets completed, e.g. if `this` is empty.

Comment (Jasper-M) [reply core/shared/src/main/scala/fs2/Stream.scala:1555]: I'm not sure. An empty stream is also an option. That's what the original implementation did afaict.

Comment (armanbilge) [reply core/shared/src/main/scala/fs2/Stream.scala:1555]: Ah I see, we are not concerned with the case `Stream.never.hold1` (which never emits but never ends) but with the case `Stream.empty.hold1`. So there should be an `onFinalize` or something.

Comment (armanbilge) [reply core/shared/src/main/scala/fs2/Stream.scala:1555]: Thanks, the fix was even easier than I realized, no need for `onFinalize` :)

**Pull request merged**